src/java/org/apache/cassandra/io/tries/TrieNode.java

```diff
@@ -0,0 +1,968 @@
+            for (int i = 0; i < childCount; ++i)
+            {
+                int next = node.transition(i);
+                while (l < next)
+                {
+                    carry = write12Bits(dest, NULL_VALUE, l - start, carry);
+                    ++l;
+                }
+                long pd = node.serializedPositionDelta(i, nodePosition);
+                carry = write12Bits(dest, (int) -pd, l - start, carry);
+                ++l;
+            }
+            if (((l - start) & 1) == 1)
+                dest.writeByte(carry);
+        }
+
+        @Override
+        boolean fits(long delta)
+        {
+            return 0 <= delta && delta <= 0xFFF;
+        }
+    };
+
+    static final LongDense LONG_DENSE = new LongDense();
+    static class LongDense extends Dense
+    {
+        // byte flags
+        // byte start
+        // byte length-1
+        // length long transition targets (-1 for not present)
+        // var payload
+        LongDense()
+        {
+            super(8);
+        }
+
+        @Override
+        public long transitionDelta(ByteBuffer src, int position, int childIndex)
+        {
+            return -src.getLong(position + 3 + childIndex * 8);
+        }
+
+        @Override
+        public void writeBytes(DataOutput dest, long ofs) throws IOException
+        {
+            dest.writeLong(ofs);
+        }
+
+        @Override
+        boolean fits(long delta)
+        {
+            return true;
+        }
+    };
+
+
+    static int read12Bits(ByteBuffer src, int base, int searchIndex)
+    {
+        int word = src.getShort(base + (3 * searchIndex) / 2);
+        if ((searchIndex & 1) == 0)
+            word = (word >> 4);
+        return word & 0xFFF;
+    }
+
+    static int write12Bits(DataOutput dest, int value, int index, int carry) throws IOException
+    {
+        assert 0 <= value && value <= 0xFFF;
+        if ((index & 1) == 0)
+        {
+            dest.writeByte(value >> 4);
+            return value << 4;
+        }
+        else
+        {
+            dest.writeByte(carry | (value >> 8));
+            dest.writeByte(value);
+            return 0;
+        }
+    }
+
+    long readBytes(ByteBuffer src, int position)
+    {
+        return SizedInts.readUnsigned(src, position, bytesPerPointer);
+    }
+
+    void writeBytes(DataOutput dest, long ofs) throws IOException
+    {
+        assert fits(ofs);
+        SizedInts.write(dest, ofs, bytesPerPointer);
+    }
+
+    boolean fits(long delta)
+    {
+        return 0 <= delta && delta < (1L << (bytesPerPointer * 8));
+    }
+
+    @Override
+    public String toString()
+    {
+        String res = getClass().getSimpleName();
+        if (bytesPerPointer >= 1)
+            res += (bytesPerPointer * 8);
+        return res;
+    }
+
+    public static Object nodeTypeString(int ordinal)
+    {
+        return values[ordinal].toString();
+    }
+
+    static final TrieNode[] values = new TrieNode[] { PAYLOAD_ONLY,
```

Comment:
This block from here to the end of the file has a handful of static classloading warnings if you run through static analysis. For example, on 951...

> Referencing subclass LongDense from superclass TrieNode initializer might lead to class loading deadlock 
 Inspection info: Reports classes that refer to their own subclasses in their static initializers or in static fields. Such references can cause JVM-level deadlocks in multithreaded environment, when one thread tries to load superclass and another thread tries to load subclass at the same time.

I would do two three things here:
1.) Move these things (`values`, `singles`, `EMPTY`, etc.) to their own class, let's say `TrieNodes`.
2.) All caps for the `static final` fields.
3.) Initialize `values`/`VALUES` in a single static method, so it's initialization isn't declaration, followed by a bunch of other stuff, followed by setting the ordinals. (i.e. Declare `singles` et al. first.)

Thoughts?

See https://www.farside.org.uk/201510/deadlocks_in_java_class_initialisation